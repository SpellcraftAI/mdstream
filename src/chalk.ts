import ansiStyles, { type ColorName, type ModifierName } from "ansi-styles"
import supportsColor from "supports-color"

// export type ColorSupportLevel = 0 | 1 | 2 | 3;

export enum ColorSupportLevel {
  None = 0,
  Colors16 = 1,
  Colors256 = 2,
  TrueColor = 3,
}

function getColorLevel(): ColorSupportLevel {
  const support = supportsColor.stdout
  if (support && support.has16m) return 3
  if (support && support.has256) return 2
  if (support && support.hasBasic) return 1
  return 0
}


/**
 * The current color level support for the process.
 */
let level: ColorSupportLevel = getColorLevel()

/**
 * Change the color level support for the process.
 */
export const setColorLevel = (newLevel?: ColorSupportLevel): ColorSupportLevel => 
  level = newLevel ?? getColorLevel()

export type RGB = [number, number, number];
export type AnsiStyle = (ModifierName | ColorName) & string;

export interface ChalkOptions {
  level?: ColorSupportLevel;
  color?: string | RGB;
  bgColor?: string | RGB;
}

function applyColor(color: string | RGB, isBackground: boolean, colorLevel: ColorSupportLevel = level): string {
  const colorFunc = isBackground ? ansiStyles.bgColor : ansiStyles.color

  if (typeof color === "string") {
    // Assume it's a hex color
    if (colorLevel === 3) return colorFunc.ansi16m(...ansiStyles.hexToRgb(color))
    if (colorLevel === 2) return colorFunc.ansi256(ansiStyles.hexToAnsi256(color))
    return colorFunc.ansi(ansiStyles.hexToAnsi(color))
  } else {
    // It's an RGB array
    if (colorLevel === 3) return colorFunc.ansi16m(...color)
    if (colorLevel === 2) return colorFunc.ansi256(ansiStyles.rgbToAnsi256(...color))
    return colorFunc.ansi(ansiStyles.rgbToAnsi(...color))
  }
}

export function chalk(text: string, styles: AnsiStyle[] = [], options: ChalkOptions = { level }): string {
  const colorLevel = options?.level ?? level

  if (colorLevel === 0) return text

  let openTags = styles.map(style => ansiStyles[style].open).join("")
  let closeTags = styles.map(style => ansiStyles[style].close).reverse().join("")

  if (options?.color) {
    openTags += applyColor(options.color, false, colorLevel)
    closeTags = ansiStyles.color.close + closeTags
  }

  if (options?.bgColor) {
    openTags += applyColor(options.bgColor, true, colorLevel)
    closeTags = ansiStyles.bgColor.close + closeTags
  }

  return `${openTags}${text}${closeTags}`
}

export type AnsiPair = { open: string; close: string };

export function getStyleTags(styles: AnsiStyle[], options: ChalkOptions = { level }): AnsiPair {
  const colorLevel = options?.level ?? level

  if (colorLevel === 0) {
    return { open: "", close: "" }
  }

  let openTags = styles.map(style => ansiStyles[style].open).join("")
  let closeTags = styles.map(style => ansiStyles[style].close).reverse().join("")

  if (options?.color) {
    openTags += applyColor(options.color, false, colorLevel)
    closeTags = ansiStyles.color.close + closeTags
  }

  if (options?.bgColor) {
    openTags += applyColor(options.bgColor, true, colorLevel)
    closeTags = ansiStyles.bgColor.close + closeTags
  }

  return { open: openTags, close: closeTags }
}