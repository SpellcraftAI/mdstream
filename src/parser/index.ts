export type * from "./types"

export { parse } from "./parse"
export { createParser, finish } from "./lib"