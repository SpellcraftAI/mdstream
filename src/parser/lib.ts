import { Token } from "../tokens"
import type { Renderer } from "../renderer"

import { parse } from "./parse"
import type { Parser } from "./types"

const TOKEN_ARRAY_CAP = 24

export function createParser<T>(renderer: Renderer<T>): Parser<T> {
  const tokens = new Uint32Array(TOKEN_ARRAY_CAP)
  tokens[0] = Token.DOCUMENT
  return {
    renderer  : renderer,
    text      : "",
    pending   : "",
    tokens    : tokens,
    len       : 0,
    token     : Token.DOCUMENT,
    codeFenceBody: 0,
    blockquoteIndex: 0,
    hrChar   : "",
    hrChars  : 0,
    backticksCount: 0,
    spaces    : new Uint8Array(TOKEN_ARRAY_CAP),
    indent    : "",
    indentLength: 0,
  }
}


/**
 * Finish rendering the markdown - flushes any remaining text.
 */
export function finish<T>(parser: Parser<T>): void {
  if (parser.pending.length > 0) {
    parse(parser, "\n")
  }
}

export function addText<T>(parser: Parser<T>): void {
  if (parser.text.length === 0) return
  // console.assert(parser.len > 0, "Never adding text to root")
  parser.renderer.addText(parser.renderer.data, parser.text)
  parser.text = ""
}

export function endToken<T>(parser: Parser<T>): void {
  // console.assert(parser.len > 0, "No nodes to end")
  parser.len -= 1
  parser.token = parser.tokens[parser.len]
  parser.renderer.endToken(parser.renderer.data, parser.tokens[parser.len + 1])
}

export function addToken<T>(parser: Parser<T>, token: Token): void {
  parser.len += 1
  parser.tokens[parser.len] = token
  parser.token = token
  parser.renderer.addToken(parser.renderer.data, token)
}

export function indexOfToken<T>(parser: Parser<T>, token: Token, startIndex: number): number {
  while (startIndex <= parser.len) {
    if (parser.tokens[startIndex] & token) {
      return startIndex
    }
    startIndex += 1
  }
  return -1
}

/**
 * End tokens until the parser has the given length.
 */
export function endTokensUntilLength<T>(parser: Parser<T>, len: number): void {
  while (parser.len > len) {
    endToken(parser)
  }
}

export function continueOrAddList<T>(parser: Parser<T>, listToken: Token): void {
  /* will create a new list inside the last item
	   if the amount of spaces is greater than the last one (with prefix)
	   1. foo
	      - bar      <- new nested ul
	         - baz   <- new nested ul
	      12. qux    <- cannot be nested in "baz" or "bar",
	                    so it's a new list in "foo"
	*/
  let listIndex = -1
  let itemIndex = -1

  for (let i = parser.blockquoteIndex + 1; i <= parser.len; i++) {
    if (parser.tokens[i] & Token.LIST_ITEM) {
      if (parser.tokens[i-1] & listToken) {
        listIndex = i-1
      }

      if (parser.indentLength < parser.spaces[i]) {
        itemIndex = -1
        break
      }

      itemIndex = i
    }
  }

  if (itemIndex === -1) {
    if (listIndex === -1) {
      endTokensUntilLength(parser, parser.blockquoteIndex)
      addToken(parser, listToken)
    } else {
      endTokensUntilLength(parser, listIndex)
    }
  } else {
    endTokensUntilLength(parser, itemIndex)
    addToken(parser, listToken)
  }
}

/**
 * Create a new list
 * or continue the last one
 */
export function addListItem<T>(parser: Parser<T>, prefixLength: number): void {
  addToken(parser, Token.LIST_ITEM)
  parser.spaces[parser.len] = parser.indentLength + prefixLength
  clearRootPending(parser)
  parser.token = Token.MAYBE_TASK
}

export function clearRootPending<T>(parser: Parser<T>): void {
  parser.indent = ""
  parser.indentLength = 0
  parser.pending = ""
}