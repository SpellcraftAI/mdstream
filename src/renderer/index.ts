export * from "./ansi"
export * from "./dom"
export * from "./html"
export * from "./log"

export type * from "./types"