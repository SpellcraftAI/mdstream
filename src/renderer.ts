import { Token, Attr } from "./tokens"
import { attr_to_html_attr } from "./parser"

export type Children<T = string> = Array<T>
export type NodeAttrs = {[key in Attr]?: string};

// Renderer types
export type RendererAddToken<T> = (data: T, type: Token) => void;
export type RendererEndToken<T> = (data: T) => void;
export type RendererAddText<T> = (data: T, text: string) => void;
export type RendererSetAttr<T> = (data: T, type: Attr, value: string) => void;

export interface Renderer<T> {
    data: T;
    add_token: RendererAddToken<T>;
    end_token: RendererEndToken<T>;
    add_text: RendererAddText<T>;
    set_attr: RendererSetAttr<T>;
}

// Default Renderer types
export interface DefaultRendererData {
  nodes: Array<HTMLElement | undefined>;
  index: number;
}

export type DefaultRenderer = Renderer<DefaultRendererData>;
export type DefaultAddToken = RendererAddToken<DefaultRendererData>;
export type DefaultEndToken = RendererEndToken<DefaultRendererData>;
export type DefaultAddText = RendererAddText<DefaultRendererData>;
export type DefaultSetAttr = RendererSetAttr<DefaultRendererData>;

export function default_renderer(root: HTMLElement): DefaultRenderer {
  return {
    add_token: default_add_token,
    end_token: default_end_token,
    add_text:  default_add_text,
    set_attr:  default_set_attr,
    data    : {
      nodes: ([root,,,,,]),
      index: 0,
    },
  }
}

export const default_add_token: DefaultAddToken = (data, type) => {
  let mount: HTMLElement
  let slot: HTMLElement

  switch (type) {
  case Token.DOCUMENT: 
    return // document is provided
  case Token.BLOCKQUOTE:
    mount = slot = document.createElement("blockquote")
    break
  case Token.PARAGRAPH:
    mount = slot = document.createElement("p")
    break
  case Token.LINE_BREAK:
    mount = slot = document.createElement("br")
    break
  case Token.RULE:
    mount = slot = document.createElement("hr")
    break
  case Token.HEADING_1:
    mount = slot = document.createElement("h1")
    break
  case Token.HEADING_2:
    mount = slot = document.createElement("h2")
    break
  case Token.HEADING_3:
    mount = slot = document.createElement("h3")
    break
  case Token.HEADING_4:
    mount = slot = document.createElement("h4")
    break
  case Token.HEADING_5:
    mount = slot = document.createElement("h5")
    break
  case Token.HEADING_6:
    mount = slot = document.createElement("h6")
    break
  case Token.ITALIC_AST:
  case Token.ITALIC_UND:
    mount = slot = document.createElement("em")
    break
  case Token.STRONG_AST:
  case Token.STRONG_UND:
    mount = slot = document.createElement("strong")
    break
  case Token.STRIKE:
    mount = slot = document.createElement("s")
    break
  case Token.CODE_INLINE:
    mount = slot = document.createElement("code")
    break
  case Token.RAW_URL:
  case Token.LINK:
    mount = slot = document.createElement("a")
    break
  case Token.IMAGE:
    mount = slot = document.createElement("img")
    break
  case Token.LIST_UNORDERED:
    mount = slot = document.createElement("ul")
    break
  case Token.LIST_ORDERED:
    mount = slot = document.createElement("ol")
    break
  case Token.LIST_ITEM:
    mount = slot = document.createElement("li")
    break
  case Token.CHECKBOX:
    const checkbox = document.createElement("input")
    checkbox.type = "checkbox"
    checkbox.disabled = true
    mount = slot = checkbox
    break
  case Token.CODE_BLOCK:
  case Token.CODE_FENCE:
    mount = document.createElement("pre")
    slot  = document.createElement("code")
    mount.appendChild(slot)
    break

  default:
    throw new Error("Unknown token: " + type)
  }

  data.nodes[data.index]?.appendChild(mount)
  data.index += 1
  data.nodes[data.index] = slot
}

export const default_end_token: DefaultEndToken = (data) => {
  data.index -= 1
}

export const default_add_text: DefaultAddText = (data, text) => {
  data.nodes[data.index]?.appendChild(document.createTextNode(text))
}

export const default_set_attr: DefaultSetAttr = (data, type, value) => {
  data.nodes[data.index]?.setAttribute(attr_to_html_attr(type), value)
}
