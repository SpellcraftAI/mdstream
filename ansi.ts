// import { createLogRenderer } from "@/renderer/log"
import { createParser, endParser, writeToParser } from "@/parser"
import { readFile } from "fs/promises"
import { createANSIRenderer } from "@/renderer/ansi"

const source = await readFile("readme.md", "utf8")

// const renderer = createLogRenderer()
const renderer = createANSIRenderer()
const parser = createParser(renderer)

let i = 0
while (i < source.length) {
  const length = 16
  const delay = Math.floor(Math.random() * 20) + 4
  const chunk = source.slice(i, i += length)
  await new Promise(resolve => setTimeout(resolve, delay))
  writeToParser(parser, chunk)
}

endParser(parser)

console.log(parser)
console.log(parser.renderer.data.buffer)