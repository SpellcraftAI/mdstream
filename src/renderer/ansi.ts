import type { Renderer } from "../types"
import { Token } from "../tokens"
import { createParser } from "../parser"
import { labelToken, MarkdownStream } from "../renderer"
import { getStyleTags, setColorLevel, type AnsiPair, type AnsiStyle, type ColorSupportLevel } from "../chalk"
import { Padding } from "../utils/PaddingStream"

const ANSI_STYLES: Partial<Record<Token, AnsiStyle[]>> = {
  [Token.DOCUMENT]: [],
  [Token.PARAGRAPH]: [],
  [Token.HEADING_1]: ["bold"],
  [Token.HEADING_2]: ["bold"],
  [Token.HEADING_3]: ["bold"],
  [Token.HEADING_4]: ["bold"],
  [Token.HEADING_5]: ["bold"],
  [Token.HEADING_6]: ["bold"],
  [Token.BLOCKQUOTE]: ["dim"],
  [Token.CODE_INLINE]: ["inverse"],
  [Token.CODE_BLOCK]: ["inverse"],
  [Token.CODE_FENCE]: ["inverse"],
  [Token.LIST_UNORDERED]: [],
  [Token.LIST_ORDERED]: [],
  [Token.LIST_ITEM]: [],
  [Token.STRONG_AST]: ["bold"],
  [Token.STRONG_UND]: ["bold"],
  [Token.ITALIC_AST]: ["italic"],
  [Token.ITALIC_UND]: ["italic"],
  [Token.STRIKE]: ["strikethrough"],
  [Token.LINK]: ["blue", "underline"],
  [Token.RAW_URL]: ["blue", "underline"],
  [Token.IMAGE]: [],
  [Token.RULE]: ["dim"],
  [Token.LINE_BREAK]: [],
  [Token.CHECKBOX]: [],
} as const

export interface ANSIRendererOptions {
  level?: ColorSupportLevel;
  render?: (chunk: string) => void;
}

const wrapLinesWithStyles = (text: string, activeStyles: AnsiPair[]) => {
  const lines = text.split("\n")
  if (lines.length > 1) {
    const openTags = activeStyles.map(({ open }) => open).join("")
    text = lines.join("\n" + openTags)
  }

  return text
}

export function createANSIRenderer({ render, level }: ANSIRendererOptions = {}): Renderer<undefined> {
  setColorLevel(level)

  let firstToken = true
  let listLevel = 0
  let prefix = ""
  let suffix = ""

  let padding: Padding
  const activeStyles: AnsiPair[] = []

  return {
    addToken: (_, token) => {
      // if (process.env.NODE_ENV !== "production") {
      //   console.log(styleText("dim", "ADDTOKEN"), labelToken(token))
      // }

      const prefixedNewline = firstToken ? "" : "\n"
      let startingNewlines = ""

      prefix = ""
      suffix = ""
      firstToken = false
      
      switch (token) {
      case Token.DOCUMENT:
        break
      case Token.PARAGRAPH:
        startingNewlines = prefixedNewline.repeat(2)
        break
      case Token.HEADING_1:
      case Token.HEADING_2:
      case Token.HEADING_3:
      case Token.HEADING_4:
      case Token.HEADING_5:
      case Token.HEADING_6:
        startingNewlines = prefixedNewline.repeat(2)
        break
      case Token.BLOCKQUOTE:
        break
      case Token.CODE_INLINE:
        prefix = " "
        suffix = " "
        break
      case Token.CODE_BLOCK:
      case Token.CODE_FENCE:
        /**
         * Create a new padding stream for this block.
         */
        padding = new Padding(1, 1, " ")
        startingNewlines = prefixedNewline.repeat(2)
        break
      case Token.LIST_UNORDERED:
      case Token.LIST_ORDERED:
        startingNewlines = listLevel < 1 ? prefixedNewline : ""
        listLevel += 1
        break
      case Token.LIST_ITEM:
        startingNewlines = prefixedNewline
        prefix = " ".repeat(listLevel * 2) + "• "
        break
      case Token.STRONG_AST:
      case Token.STRONG_UND:
        break
      case Token.ITALIC_AST:
      case Token.ITALIC_UND:
        break
      case Token.STRIKE:
        break
      case Token.LINK:
      case Token.RAW_URL:
        break
      case Token.IMAGE:
        prefix = "[Image] "
        break
      case Token.RULE:
        startingNewlines = prefixedNewline.repeat(2)
        prefix =  "─".repeat(40)
        break
      case Token.LINE_BREAK:
        startingNewlines = prefixedNewline
        break
      case Token.CHECKBOX:
        prefix = "☐" // "✅"
        break
      }

      render?.(startingNewlines)
      const styles = ANSI_STYLES[token]
      if (styles) {
        const tags = getStyleTags(styles)
        render?.(tags.open)
        activeStyles.push(tags)
      }
      render?.(prefix)
    },
    endToken: (_, token) => {
      // if (process.env.NODE_ENV !== "production") {
      //   console.log(styleText("dim", "ENDTOKEN"), labelToken(token))
      // }
      switch (token) {
      case Token.LIST_ORDERED:
      case Token.LIST_UNORDERED:
        listLevel -= 1
        break

      case Token.CODE_BLOCK:
      case Token.CODE_FENCE:
        render?.(padding.flush())
        break
      }

      render?.(suffix)
      const styles = ANSI_STYLES[token]
      if (styles) {
        const tags = getStyleTags(styles)
        activeStyles.pop()
        render?.(tags.close)
      }
    },
    addText: (_, token, text) => {
      // if (process.env.NODE_ENV !== "production") {
      //   console.log(styleText("dim", "ADDTEXT"), labelToken(token), JSON.stringify(text))
      // }
      /**
       * For multiline text, we will make sure to end the ANSI sequence at the
       * end of the line, and re-start it at the beginning of the next line.
       * 
       * This is to prevent boxes or other transforms that work with newlines
       * from interrupting the multiline styling, so it's restarted each line.
       */
      // const lines = text.split("\n")
      // if (lines.length > 1) {
      //   const openTags = activeStyles.map(({ open }) => open).join("")
      //   text = lines.join("\n" + openTags)
      // }

      switch (token) {
      case Token.CODE_BLOCK:
      case Token.CODE_FENCE:
        text = padding.processChunk(text)
        // render?.(wrapLinesWithStyles(padded, activeStyles))
        break
      }

      const withStyledLines = wrapLinesWithStyles(text, activeStyles)
      render?.(withStyledLines)
    },
    setAttr: () => {
    },
    data: undefined,
  }
}


export class MarkdownANSIStream extends MarkdownStream<undefined> {
  constructor(level?: ColorSupportLevel) {
    const ENCODER = new TextEncoder()
    
    super({
      start: (controller) => {
        const renderer = createANSIRenderer({
          level,
          render: (chunk) => controller.enqueue(ENCODER.encode(chunk)),
        })

        return createParser(renderer)
      }
    })
  }
}