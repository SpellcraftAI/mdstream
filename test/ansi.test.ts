import { expect, test, describe, afterAll } from "bun:test"
import { createParser, parse, finish } from "@/parser"
import { createANSIRenderer, MarkdownANSIStream } from "@/renderer/ansi"
import { readFile } from "fs/promises"
import chalk from "chalk"

const originalChalkLevel = chalk.level
const encoder = new TextEncoder()

describe("Streaming Markdown Parser", () => {
  const testParser = async () => {
    let source = await readFile("readme.md", "utf8")
    const renderer = createANSIRenderer({
      render: (chunk) => process.stdout.write(chunk),
    })

    const parser = createParser(renderer)

    let i = 0
    while (i < source.length) {
      const length = 16
      const chunk = source.slice(i, i += length)
      parse(parser, chunk)
    }

    finish(parser)
    return parser.renderer.data.buffer
  }

  test("should correctly parse and render markdown with color", async () => {
    chalk.level = 1
    const result = await testParser()
    expect(encoder.encode(result)).toMatchSnapshot("ansi-force-color")
  })

  test("should correctly parse and render markdown without color", async () => {
    chalk.level = 0
    const result = await testParser()
    expect(encoder.encode(result)).toMatchSnapshot("ansi-force-no-color")
  })

  const testStream = async () => {
    const decoder = new TextDecoder()
    const file = Bun.file("readme.md")
    const stream = file.stream()

    let result = ""
    const outputStream = new WritableStream({
      write(chunk) {
        result += decoder.decode(chunk)
      }
    })

    await stream.pipeThrough(new MarkdownANSIStream()).pipeTo(outputStream)
    return result
  }

  test("should correctly process markdown through MarkdownANSIStream", async () => {
    chalk.level = 1
    const result = await testStream()
    expect(encoder.encode(result)).toMatchSnapshot("markdown-ansi-stream-color")
  })

  test("should correctly process markdown through MarkdownANSIStream", async () => {
    chalk.level = 0
    const result = await testStream()
    expect(encoder.encode(result)).toMatchSnapshot("markdown-ansi-stream-no-color")
  })

  afterAll(() => {
    chalk.level = originalChalkLevel
  })
})