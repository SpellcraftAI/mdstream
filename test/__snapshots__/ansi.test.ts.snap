// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Streaming Markdown Parser should correctly parse and render markdown with color: ansi-force-color 1`] = `
"

[1mStreaming [22m[3m[1mMa[22m[23m[3m[1mrkdown[22m[23m

[35m[Image] [39m[34m[4mvers[24m[39m[34m[4mion[24m[39m[34m (https://img.shields.io/npm/v/streaming-markdown?logo=npm)[39m[34m (https://www.npmjs.com/package/streaming-markdown)[39m [35m[Image] [39m[34m[4mgithub[24m[39m[34m (https://img.shields.io/badge/GitHub-streaming--markdown-orange?logo=github)[39m[34m (https://github.com/thetarnav/streaming-markdown)[39m

[1mEx[22m[1mperiment making [22m[1ma streaming mark[22m[1mdown parser [22m[3m[1mÃ  l[22m[23m[3m[1ma ChatGPT.[22m[23m[2m[22m
[2m[22m
[2mâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€[22m

[1mInstallat[22m[1mion[22m

Install [100m[37mstr[39m[49m[100m[37meaming-markdown[39m[49m package from npm.

[100m[37mnpm [39m[49m[100m[37minstall streamin[39m[49m[100m[37mg-markdown[39m[49m

[3mOr just copy [23m[100m[37m[1m[34m[4m[3msmd.js[23m[24m[39m[37m[22m[39m[49m[34m (https://github.com/thetarnav/streaming-markdown/blob/main/smd.js)[39m file to your project.

Or use the [34m[4mCDN link[24m[39m[34m (https://www.jsdelivr.com/package/npm/streaming-markdown)[39m.

It's a minified [3m(3kB G[23m[3mzip)[23m version of the package, with only the necessary functions exported.

See the exports in [100m[37m[34m[4ms[24m[39m[37m[39m[49m[100m[37m[34m[4mmd_min_entry.js[24m[39m[37m[39m[49m[34m (https://github.com/thetarnav/streaming-markdown/blob/main/smd_min_entry.js)[39m.

The package uses ES module exports, so you need to use [100m[37mtype="mo[39m[49m[100m[37mdule"[39m[49m in your script tag.

[100m[37m<script type[39m[49m[100m[37m="module">[39m[49m
[100m[37m    i[39m[49m[100m[37mmport * as smd f[39m[49m[100m[37mrom "https://cdn[39m[49m[100m[37m.jsdelivr.net/np[39m[49m[100m[37mm/streaming-mark[39m[49m[100m[37mdown/smd.min.js"[39m[49m[100m[37m[39m[49m
[100m[37m    // ...[39m[49m
[100m[37m</sc[39m[49m[100m[37mript>[39m[49m

[1mU[22m[1msage[22m

First create new markdown [100m[37mParser[39m[49m by calling [100m[37mparser[39m[49m function.

It's single argument is a [100m[37mRenderer[39m[49m object, which is an interface to render the parsed markdown tokens to the DOM.

There are two built-in renderersâ€”[100m[37mde[39m[49m[100m[37mfault_renderer[39m[49m and [100m[37mlogger_rend[39m[49m[100m[37merer[39m[49mâ€”that you can try at first.

[100m[37mimport *[39m[49m[100m[37m as smd from "st[39m[49m[100m[37mreaming-markdown[39m[49m[100m[37m"[39m[49m
[100m[37m[39m[49m
[100m[37mconst element[39m[49m[100m[37m  = document.get[39m[49m[100m[37mElementById("mar[39m[49m[100m[37mkdown")[39m[49m
[100m[37mconst re[39m[49m[100m[37mnderer = smd.def[39m[49m[100m[37mault_renderer(el[39m[49m[100m[37mement)[39m[49m
[100m[37mconst par[39m[49m[100m[37mser   = smd.pars[39m[49m[100m[37mer(renderer)[39m[49m

[100m[37m[1mwrite[22m[39m[49m function

Then, you can start streaming markdown to the [100m[37mParser[39m[49m by calling [100m[37mparse[39m[49m[100m[37mr_write[39m[49m function with the chunk of markdown string.

[100m[37msmd.[39m[49m[100m[37mparser_write(par[39m[49m[100m[37mser, "# Streamin[39m[49m[100m[37mg Markdo[39m[49m"
`;

exports[`Streaming Markdown Parser should correctly parse and render markdown without color: ansi-force-no-color 1`] = `
"

Streaming Markdown

[Image] version (https://img.shields.io/npm/v/streaming-markdown?logo=npm) (https://www.npmjs.com/package/streaming-markdown) [Image] github (https://img.shields.io/badge/GitHub-streaming--markdown-orange?logo=github) (https://github.com/thetarnav/streaming-markdown)

Experiment making a streaming markdown parser Ã  la ChatGPT.

â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€

Installation

Install streaming-markdown package from npm.

npm install streaming-markdown

Or just copy smd.js (https://github.com/thetarnav/streaming-markdown/blob/main/smd.js) file to your project.

Or use the CDN link (https://www.jsdelivr.com/package/npm/streaming-markdown).

It's a minified (3kB Gzip) version of the package, with only the necessary functions exported.

See the exports in smd_min_entry.js (https://github.com/thetarnav/streaming-markdown/blob/main/smd_min_entry.js).

The package uses ES module exports, so you need to use type="module" in your script tag.

<script type="module">
    import * as smd from "https://cdn.jsdelivr.net/npm/streaming-markdown/smd.min.js"
    // ...
</script>

Usage

First create new markdown Parser by calling parser function.

It's single argument is a Renderer object, which is an interface to render the parsed markdown tokens to the DOM.

There are two built-in renderersâ€”default_renderer and logger_rendererâ€”that you can try at first.

import * as smd from "streaming-markdown"

const element  = document.getElementById("markdown")
const renderer = smd.default_renderer(element)
const parser   = smd.parser(renderer)

write function

Then, you can start streaming markdown to the Parser by calling parser_write function with the chunk of markdown string.

smd.parser_write(parser, "# Streaming Markdo"
`;

exports[`Streaming Markdown Parser should correctly process markdown through MarkdownANSIStream: markdown-ansi-stream-color 1`] = `
"

[1mStreaming [22m[3m[1mMarkdown[22m[23m

[35m[Image] [39m[34m[4mversion[24m[39m [35m[Image] [39m[34m[4mgithub[24m[39m

[1mExperiment making a streaming markdown parser [22m[3m[1mÃ  la ChatGPT.[22m[23m[2m[22m
[2m[22m
[2mâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€[22m

[1mInstallation[22m

Install [100m[37mstreaming-markdown[39m[49m package from npm.

[100m[37mnpm install streaming-markdown[39m[49m

[3mOr just copy [23m[100m[37m[1m[34m[4m[3msmd.js[23m[24m[39m[37m[22m[39m[49m file to your project.

Or use the [34m[4mCDN link[24m[39m.

It's a minified [3m(3kB Gzip)[23m version of the package, with only the necessary functions exported.

See the exports in [100m[37m[34m[4msmd_min_entry.js[24m[39m[37m[39m[49m.

The package uses ES module exports, so you need to use [100m[37mtype="module"[39m[49m in your script tag.

[100m[37m<script type="module">[39m[49m
[100m[37m    import * as smd from "https://cdn.jsdelivr.net/npm/streaming-markdown/smd.min.js"[39m[49m
[100m[37m    // ...[39m[49m
[100m[37m</script>[39m[49m

[1mUsage[22m

First create new markdown [100m[37mParser[39m[49m by calling [100m[37mparser[39m[49m"
`;

exports[`Streaming Markdown Parser should correctly process markdown through MarkdownANSIStream: markdown-ansi-stream-no-color 1`] = `
"

Streaming Markdown

[Image] version [Image] github

Experiment making a streaming markdown parser Ã  la ChatGPT.

â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€

Installation

Install streaming-markdown package from npm.

npm install streaming-markdown

Or just copy smd.js file to your project.

Or use the CDN link.

It's a minified (3kB Gzip) version of the package, with only the necessary functions exported.

See the exports in smd_min_entry.js.

The package uses ES module exports, so you need to use type="module" in your script tag.

<script type="module">
    import * as smd from "https://cdn.jsdelivr.net/npm/streaming-markdown/smd.min.js"
    // ...
</script>

Usage

First create new markdown Parser by calling parser function.

It's single argument is a Renderer object, which is an interface to render the parsed markdown tokens to the DOM.

There are two built-in renderersâ€”default_renderer and logger_rendererâ€”that you can try at first.

import * as smd from "streaming-markdown"

const element  = document.getElementById("markdown")
const renderer = smd.default_renderer(element)
const parser   = smd.parser(renderer)

write function

Then, you can start streaming markdown to the Parser"
`;
